### DISCUSSION TITLE
Cannot run on parallel in a local HPC

### DISCUSSION THREAD

**Discussion by saeedvdi** [idaholab/moose#25686] [Q&A Getting Started] [answered]
Hello!
I have been using a moose based FEM package on moose without no problem; however, after i added a new auxiliary variable that calculate the plastic deformation, I get the following error whenever I am truing to run it on parallel. I was wondering if you could help me figure this out. 
```

(moose) s_vadiee@speed-32:/nfs/speed-scratch/s_vadiee/September/00-Dip45-base-case/ICs$ mpiexec -n 32 //speed-scratch/s_vadiee/opt/golem/golem-opt -i 05-THM-Plastic-Slip-3000m.i
Setting Up.
  Finished Setting Mesh                                                                  [  2.04 s] [  442 MB]
Still Setting Up...
  Initializing
    Finished Initializing Equation Systems                                               [  8.86 s] [  666 MB]
    Finished Initializing Displaced Equation System                                      [  8.20 s] [  787 MB]
  Finished Initializing                                                                  [ 19.96 s] [  788 MB]
Finished Setting Up                                                                      [ 60.79 s] [  789 MB]
Framework Information:
MOOSE Version:           git commit 0db956d593 on 2023-08-20
LibMesh Version:         
PETSc Version:           3.16.6
SLEPc Version:           3.16.2
Current Time:            Sun Oct  8 12:20:14 2023
Executable Timestamp:    Mon Aug 21 11:17:55 2023

Parallelism:
  Num Processors:          32
  Num Threads:             1

Mesh: 
  Parallel Type:           replicated
  Mesh Dimension:          3
  Spatial Dimension:       3
  Nodes:                   
    Total:                 119732
    Local:                 4344
    Min/Max/Avg:           3154/4344/3741
  Elems:                   
    Total:                 692523
    Local:                 21768
    Min/Max/Avg:           21231/22357/21641
  Num Subdomains:          4
  Num Partitions:          32
  Partitioner:             metis
onlinear System:
  Num DOFs:                598660
  Num Local DOFs:          21720
  Variables:               { "disp_x" "disp_y" "disp_z" "pore_pressure" "temperature" } 
  Finite Element Types:    "LAGRANGE" 
  Approximation Orders:    "FIRST" 

Auxiliary System:
  Num DOFs:                26639633
  Num Local DOFs:          837932
  Variables:               { "stress_xx" "stress_yy" "stress_zz" "stress_xy" "stress_xz" ... "strain_yy" 
                             "strain_zz" "strain_xy" "strain_xz" "strain_yz" } { "vf_x" "vf_y" "vf_z" 
                             "porosity" "fluid_density" "fluid_viscosity" "permeability" } { "EP" "ES_xx" 
                             "ES_yy" "ES_zz" "plastic_strain_xx" ... "plastic_strain_xz" "plastic_strain_yz" 
                             "plastic_strain_eqv" "plastic_strain_rate_eqv" "yield" } { "normal_stress" 
                             "shear_stress" "slip_tendency" } 
  Finite Element Types:    "MONOMIAL" "MONOMIAL" "MONOMIAL" "MONOMIAL" 
  Approximation Orders:    "CONSTANT" "CONSTANT" "CONSTANT" "CONSTANT" 

Execution Information:
  Executioner:             Transient
  TimeStepper:             FunctionDT
  TimeIntegrator:          ImplicitEuler
  Solver Mode:             NEWTON
  MOOSE Preconditioner:    FSP

Currently Executing
  Performing Initial Setup
    Setting Up Materials
      Finished Computing Initial Material Values                                         [  0.71 s] [  899 MB]
    Finished Setting Up Materials                                                        [  0.71 s] [  899 MB]
    Building SemiLocalElemMap....                                                        [ 31.18 s] [  938 MB]
  Finished Performing Initial Setup                                                      [ 34.08 s] [  938 MB]

Time Step 0, time = 0
ime Step 1, time = 1, dt = 1

===================================================================================
=   BAD TERMINATION OF ONE OF YOUR APPLICATION PROCESSES
=   PID 16449 RUNNING AT speed-32.encs.concordia.ca
=   EXIT CODE: 11
=   CLEANING UP REMAINING PROCESSES
=   YOU CAN IGNORE THE BELOW CLEANUP MESSAGES
===================================================================================
YOUR APPLICATION TERMINATED WITH THE EXIT STRING: Segmentation fault (signal 11)
This typically refers to a problem with your application.
Please see the FAQ page for debugging suggestions

```

**Comment by GiudGiud** [comment]
Hello

This is a genuine bug, signal 11 means we're reading out of bounds most likely
Could you please share your input with us? A git repo works best for us. 

Alternatively if you want to look for it yourself, seg faults are easy to find using a debugger
https://mooseframework.inl.gov/application_development/debugging.html

Guillaume

**Reply by saeedvdi** [reply]
Yes, here is the link to my input file

https://github.com/saeedvdi/GeoSaslatoon_Conference.git

**Reply by GiudGiud** [reply]
And this runs with Golem?

**Reply by saeedvdi** [reply]
Yes

**Reply by GiudGiud** [reply]
looks like I need your mesh to run this input.
If you cant share your mesh do you mind making a dummy mesh with the same block / boundary names. 
Using these generators for example
https://mooseframework.inl.gov/source/meshgenerators/CartesianMeshGenerator.html
https://mooseframework.inl.gov/source/meshgenerators/StitchedMeshGenerator.html
https://mooseframework.inl.gov/source/meshgenerators/RenameBoundaryGenerator.html

**Reply by saeedvdi** [reply]
I have uploaded the mesh in the same git repository. 
Thank you in advance for your assistance.

**Reply by GiudGiud** [reply]
looks to be a ghosting issue, that s what I get compiling in devel mode
```

    Computing Initial Residual
      Computing Residual.libMesh terminating:
No index 1948079 in ghosted vector.
Vector contains [3297537,6592620)
And empty ghost array.
```

EDIT: still looking for the object causing this

**Reply by GiudGiud** [reply]
seems this object and the two below are involved.
```
  [normal_stress]
    type = GolemFaultNormalStress
    variable = normal_stress
    boundary = '9'
    check_boundary_restricted = false
  []
```
Boundary restricted auxkernels can be a little harder to get right. Do people usually use this with a boundary restriction?
Internal to the mesh or external?
It looks internal in your mesh

**Reply by GiudGiud** [reply, marked as answer]
It's kind of hacky but all you need is to add a layer of ghosting to avoid the error (I havent checked the results)
To do that, just use an object that lets you add that. For example,
```
[AuxKernels]
  [add_ghosting]
    type = ProjectionAux
    variable = dummy
    v = yield
    ghost_layers = 1
    execute_on = 'NONE'
    block = 1
  []
[]

[AuxVariables]
  [dummy]
    order = CONSTANT
    family = MONOMIAL
  []
[]
```

It would be great to engage with the Golem developers to discuss why they need ghosting on these 3 boundary-restricted auxkernels. Maybe it s simply that to compute the stress, one needs the neighbor values to form a derivative

**Reply by ajacquey** [reply]
Hi,

First, these `AuxKernels` implemented in Golem are experimental. I was trying to find an easy way to output normal and shear stress to a surface (internal) based on the stress tensor in the surrounding 3D elements. I hadn't had the time to properly test these `AuxKernels` but I am glad someone tried to use them.

@GiudGiud As I said above, I was trying to output the normal and shear stress to a fault (represented as a sideset and a lower dimensional material). The main requirement is to be able to compute the traction on the surface (see [this line ](https://github.com/ajacquey/golem/blob/a77701b5882a4f89d8aa2b119c6fb18850f915c9/src/auxkernels/GolemFaultShearStress.C#L42) in `GolemFaultShearStress.C`) as the product between the stress tensor and the unit normal vector to the surface.

**Comment by Rahim-Habibi** [comment]
Hi, 
According to the response of @GiudGiud meaning that you have problem in "normal stress" block. 
Actually I am working on such problems (simulation) as well.  So I wrote auxkernels for computing stress on such condition (e.g. computing stress on fault in deep reservoir modeling). I will do "pull request" to add i them on MOOSE soon.

**Comment by saeedvdi** [comment]
@GiudGiud I was able to figure out the problem, and now I can run it in parallel. You can close the discussion. And thank you for your help
